File: matlab/run_game.m
%RUN_GAME  Single-game runner with per-turn logging and VP chart.
%
%   Runs one full game (heuristic vs mcts, no display).
%   Saves timestamped outputs to matlab/outputs/:
%     YYYYMMDD_HHMMSS_game_log.csv
%     YYYYMMDD_HHMMSS_vp_over_time.png
%
%   Run:  run_game

%% ── PARAMS ───────────────────────────────────────────────────────────────────
PARAMS.agents  = {'heuristic', 'mcts'};
PARAMS.rngSeed = 42;

%% ── Setup ────────────────────────────────────────────────────────────────────
ts        = datestr(now, 'yyyymmdd_HHMMSS');  %#ok<TNOW1,DATST>
outputDir = fullfile(fileparts(mfilename('fullpath')), 'outputs');
if ~exist(outputDir, 'dir'), mkdir(outputDir); end

config                = catan_core('defaultConfig');
config.showViz        = false;
config.pauseAfterMove = false;
config.verbose        = false;
config.rngSeed        = PARAMS.rngSeed;

agentNames = PARAMS.agents;
agentFnMap = containers.Map( ...
    {'random','heuristic','monte_carlo','mcts'}, ...
    {@agent_random, @agent_heuristic, @agent_montecarlo, @agent_mcts});

agentFns = cellfun(@(n) agentFnMap(n), agentNames, 'UniformOutput', false);
config.numPlayers = numel(agentFns);

%% ── Run game ─────────────────────────────────────────────────────────────────
history = catan_core('simulateGame', agentFns, config, agentNames);
fs      = history.finalState;

%% ── Console output ───────────────────────────────────────────────────────────
winId   = fs.winnerId;
if winId >= 1 && winId <= numel(agentNames)
    winName = agentNames{winId};
else
    winName = 'None';
end

fprintf('\nGame over!\n');
fprintf('Winner     : %s (Player %d)\n', winName, winId);
fprintf('Total turns: %d\n', fs.turnIndex);
for p = 1:numel(agentNames)
    fprintf('  P%d (%s): %d VP\n', p, agentNames{p}, fs.players(p).victoryPoints);
end

%% ── game_log.csv ─────────────────────────────────────────────────────────────
fid = fopen(fullfile(outputDir, sprintf('%s_game_log.csv', ts)), 'w');
fprintf(fid, 'turn,player_id,agent_type,action_type,vp_after\n');
for k = 1:numel(history.actions)
    e = history.actions(k);
    fprintf(fid, '%d,%d,%s,%s,%d\n', ...
        e.turn, e.player, agentNames{e.player}, e.type, e.vp);
end
fclose(fid);

%% ── vp_over_time.png ─────────────────────────────────────────────────────────
% Group actions by player to build per-agent VP time series.
% Each recorded entry is the acting player's VP after their action.
% stairs() draws a step function that stays flat between data points.

palette = [0.357 0.608 0.835;   % muted blue
           0.929 0.490 0.192];  % muted orange

fig2 = figure('Visible','off','Position',[0 0 820 420]);
ax2  = axes(fig2);
hold(ax2, 'on');

for p = 1:numel(agentNames)
    mask = [history.actions.player] == p;
    if ~any(mask), continue; end
    entries = history.actions(mask);
    turns   = [entries.turn];
    vps     = [entries.vp];
    stairs(ax2, turns, vps, ...
        'Color',    palette(mod(p-1, size(palette,1))+1, :), ...
        'LineWidth', 2, ...
        'DisplayName', agentNames{p});
end

hold(ax2, 'off');
legend(ax2, 'Location','northwest');
xlabel(ax2, 'Turn');
ylabel(ax2, 'Victory Points');
title(ax2, 'VP Progression');
ax2.YGrid         = 'on';
ax2.XGrid         = 'off';
ax2.GridLineStyle = '--';
ax2.GridAlpha     = 0.4;
ax2.TickLength    = [0, 0];
box(ax2, 'off');

print(fig2, fullfile(outputDir, sprintf('%s_vp_over_time.png', ts)), '-dpng', '-r150');
close(fig2);

fprintf('\n  Outputs saved to: %s\n', outputDir);
fprintf('  Timestamp prefix: %s\n\n', ts);

File: run_game.py
"""
run_game.py — Single-game runner with per-turn logging and VP chart.

Runs one full game (heuristic vs mcts, no display).
Saves timestamped outputs to outputs/:
  YYYYMMDD_HHMMSS_game_log.csv
  YYYYMMDD_HHMMSS_vp_over_time.png

Usage
-----
  python run_game.py
"""

import csv
import os
from collections import defaultdict
from datetime import datetime

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

from config import default_config
from agents import resolve_agent
from game import simulate_game

# ── Defaults ──────────────────────────────────────────────────────────────────

DEFAULT_AGENTS = ["heuristic", "mcts"]
OUTPUT_DIR     = "outputs"
COLORS         = ["#5B9BD5", "#ED7D31"]


# ── Runner ────────────────────────────────────────────────────────────────────

def run_single_game(
    agent_names: list = None,
    config: dict = None,
) -> dict:
    """
    Run one game and save a per-turn CSV log and VP-over-time chart.

    Parameters
    ----------
    agent_names : list of str
        Exactly 2 agent names (default: ['heuristic', 'mcts']).
    config : dict, optional
        Base config dict; verbose is forced False.

    Returns
    -------
    history dict from simulate_game
    """
    if agent_names is None:
        agent_names = DEFAULT_AGENTS
    if config is None:
        config = default_config()

    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    cfg                = dict(config)
    cfg["verbose"]     = False
    cfg["num_players"] = len(agent_names)

    agent_fns = [resolve_agent(n) for n in agent_names]
    history   = simulate_game(agent_fns, cfg, player_names=agent_names)
    state     = history["final_state"]

    # ── Console output ────────────────────────────────────────────────────────

    winner_pid  = state.winner_id
    winner_name = (agent_names[winner_pid - 1]
                   if 1 <= winner_pid <= len(agent_names)
                   else "None")

    print(f"\nGame over!")
    print(f"Winner     : {winner_name} (Player {winner_pid})")
    print(f"Total turns: {state.turn_index}")
    for pid, name in enumerate(agent_names, start=1):
        print(f"  P{pid} ({name}): {state.players[pid].victory_points} VP")

    # ── game_log.csv ──────────────────────────────────────────────────────────

    log_path = os.path.join(OUTPUT_DIR, f"{ts}_game_log.csv")
    with open(log_path, "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["turn", "player_id", "agent_type", "action_type", "vp_after"])
        for entry in history["actions"]:
            w.writerow([
                entry["turn"],
                entry["player"],
                agent_names[entry["player"] - 1],
                entry["type"],
                entry["vp"],
            ])

    # ── vp_over_time.png ──────────────────────────────────────────────────────
    # Build one (turns, vps) series per player from the action log.
    # Only turns where the player acted are recorded; ax.step() fills the gaps.

    vp_series: dict = defaultdict(list)  # pid → [(turn, vp), ...]
    for entry in history["actions"]:
        vp_series[entry["player"]].append((entry["turn"], entry["vp"]))

    fig, ax = plt.subplots(figsize=(10, 5))

    for pid, name in enumerate(agent_names, start=1):
        pts = vp_series.get(pid, [])
        if not pts:
            continue
        turns = [t for t, _ in pts]
        vps   = [v for _, v in pts]
        ax.step(turns, vps, where="post",
                label=name, color=COLORS[pid - 1], linewidth=2)

    ax.set_xlabel("Turn")
    ax.set_ylabel("Victory Points")
    ax.set_title("VP Progression")
    ax.legend()
    ax.yaxis.grid(True, linestyle="--", alpha=0.4)
    ax.xaxis.grid(False)
    ax.set_axisbelow(True)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    plt.tight_layout()

    chart_path = os.path.join(OUTPUT_DIR, f"{ts}_vp_over_time.png")
    plt.savefig(chart_path, dpi=150, bbox_inches="tight")
    plt.close()

    print(f"\n  Outputs saved to {OUTPUT_DIR}/{ts}_*.{{csv,png}}")
    return history


# ── CLI ───────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    run_single_game()
